### PR TITLE
Extension framework + Feature-based alignment

### PR DESCRIPTION
**Pull request by leongwaikay**
Quite a number of things in this PR, as I was refactoring and tidying up as I go (Just following the Boy Scouts rule of leaving your code better than you found it). Ideally each new point should be a separate PR on its own but it was too intertwined to separate when I was done with writing.

## Summary of changes
- Added feature-based alignment by using a reference image. Works wonders when there are no page boundaries or markers to realigned a scanned document.
  - See sample5
- Modularised the image preprocessing into plugins/extensions/modules. Not sure which name fits so I just called them extensions for now. Basically the template can specify what kind of preprocessing is needed to prepare the images for OMR.
  - New preprocessors can just be placed arbitrary nested in the extension directory and as long as they subclass `ImagePreprocessor` they will be loaded.
  - Added some builtin preprocessors like GaussianBlur and Levels.
  - Move markers into its own module. (Need much testing here as I do not have sheets with markers)
  - Not sure if template is the best place to specify preprocessors as it is dependent on the way the particular images are obtained. Not so much the structure of the form. Maybe we should differentiate between `template` and `configuration` file.
- Keys of `template.json` are now case insensitive and ordered. Basically, they are converted to lowercase and then placed in an OrderedDict. This is because order matters to the application of the image preprocessors. And it is confusing to keep wondering what case the keys are.
- Refactored to use `pathlib` for most filepaths. `os.path.join` is too unclean. Next step should be to clean up the output file paths.

### CODE REVIEW

**Comment by Udayraj123** [comment]
> Ideally each new point should be a separate PR on its own but it was too intertwined to separate when I was done with writing.
No problem! As long as the code is improving we won't face such issues in the future.

> Need much testing here as I do not have sheets with markers
Yeah, they were updated into sample1 folder. I will test by updating their template.json files (might need a bit help from you here)

> Not sure which name fits so I just called them extensions for now
modules would be a better term I think.

> Not sure if template is the best place to specify preprocessors
Yes, template.json shall not contain the whole preprocessing config. At-least there should not be a compulsion to supply it in every template(As it is not the user's worry). Also, the `--noMarker` flag, etc would have an alternative through this, might increase confusion on using both.

> Maybe we should differentiate between template and configuration file.
Valid point, maybe an additional config.json file which if present, will override the default config including preprocessing and the flags.

> And it is confusing to keep wondering what case the keys are.
We can follow camelCasing for all keys, lowercase would cause readability issues.

> Next step should be to clean up the output file paths.
Would look into it :)

File: extensions/CropPage.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+import cv2
+import numpy as np
+from extension import ImagePreprocessor
+
+MIN_PAGE_AREA = 80000
+
+# TODO: (remove noCropping bool) Automate the case of close up scan(incorrect page)-
+# ^Note: App rejects croppeds along with others
+def normalize(image):
```
Comment (Udayraj123) [inline extensions/CropPage.py]: Can further modularize this snippet. Perhaps in the future.

File: main.py [hunk 2]
```diff
@@ -301,59 +320,57 @@ def process_files(omr_files, template, args, out):
...
         if(args["setLayout"]):
             templateLayout = utils.drawTemplateLayout(
-                OMRCrop, template, shifted=False, border=2)
+                inOMR, template, shifted=False, border=2)
             utils.show("Template Layout", templateLayout, 1, 1)
             continue
 
         # uniquify
-        file_id = inputFolderName + '_' + filename
+        file_id = str(filename)
```
Comment (Udayraj123) [inline main.py]: filename may not be unique across directories

**Comment by Udayraj123** [comment]
This branch would contain my changes(conflict resolutions and sample case fixes etc) over your commits for merging this PR with conflicts-
https://github.com/Udayraj123/OMRChecker/tree/leongwaikay-extension-framework

File: extensions/Markers.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        if(not os.path.exists(self.marker_path)):
+            print(
+                "Error: Marker not found at path provided in template:",
+                self.marker_path)
+            exit(31)
+
+        marker = cv2.imread(self.marker_path, cv2.IMREAD_GRAYSCALE)
+
+        if("sheettomarkerwidthratio" in marker_ops):
```
[the remark this answers is not in the document]
Comment (leongwaikay) [reply extensions/Markers.py]: Basically we want the keys in the json to be case insensitive. It is not easy for non-programmers to understand camelCase.

I do not think code readability will be a big issue as there are not a lot of options. But there are two solutions:
1. do something like `'SheetToMarkerWidthRatio'.lower() in marker_ops`
2. Further modify the dictionary to match keys disregarding case.

File: main.py (hunk 2 above)
Comment (leongwaikay) [reply main.py]: The output now follows the same directory structure as the input. So there will not be an issue with files of same names as they will be in their own directory

**Comment by Udayraj123** [comment]
@leongwaikay just giving update - I will soon find some time to finish reviewing this PR. Sorry to keep you waiting a bit further 😅

**Comment by Udayraj123** [comment]
@leongwaikay I am currently testing out my samples with the new configurations. Specifically, using the Markers.py extension(now renamed to CropOnMarkers.py) with sample1. 
I see that `getBestMatch(...)` from utils.py is not yet ready to accept the new ImageProcessor. 
```
Exception has occurred: AttributeError
module 'config' has no attribute 'marker_rescale_range'
  File "/OMRChecker/utils.py", line 267, in getBestMatch
    config.marker_rescale_range[1] - config.marker_rescale_range[0]) // config.marker_rescale_steps
  File "/OMRChecker/extensions/CropOnMarkers.py", line 78, in apply_filter
    best_scale, allMaxT = utils.getBestMatch(image_eroded_sub, self.marker)
  File "/OMRChecker/main.py", line 332, in process_files
    inOMR = pp.apply_filter(inOMR, args)
  File "/OMRChecker/main.py", line 85, in process_dir
    process_files(omr_files, template, args_local, output_set)
  File "/OMRChecker/main.py", line 93, in process_dir
    process_dir(root_dir, d, template)
  File "/OMRChecker/main.py", line 93, in process_dir
    process_dir(root_dir, d, template)
  File "/OMRChecker/main.py", line 582, in <module>
    process_dir(Path(root), Path(root), args['template'])
```
Will fix this so that Older samples also get compatible with this configuration

**Comment by Udayraj123** [comment]
@leongwaikay I have made changes in your branch to make it compatible with my samples - can you merge them(https://github.com/leongwaikay/OMRChecker/pull/2 ) into your branch first?

File: extensions/Markers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Udayraj123) [reply extensions/Markers.py]: lower case keys will work for now as we can still use camel case in template.json file.

**Pull request merged**